DSDEGP-1261: clio integration test in jenkins

pshapiro4broad opened the pull request:
### Description

DSDEGP-1261: modify clio integration test to run on jenkins and to use cloud elastic search (not docker elastic search)

This change adds a new sbt task `testDockerJenkins` which runs the same tests as `testDocker` but uses `elasticsearch1.gotc-dev.broadinstitute.org` instead of the docker elastic search host.

I've tested this out by configuring a Jenkins job to build this new task, the output of that build is here: https://gotc-jenkins.dsp-techops.broadinstitute.org/job/clio-integration/11/console

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [x] **Submitter**: Verify all tests go green, including CI tests
- [x] **Submitter**: Squash commits and merge to develop
- [x] **Submitter**: Delete branch after merge
- [x] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [x] **Submitter**: Move the JIRA issue to QA once this checklist is completed

pshapiro4broad committed:
DSDEGP-1261: integration tests run on jenkins, implemented as separate tests

pshapiro4broad committed:
DSDEGP-1261: implement jenkins integration test

- combine jenkins test code with existing test code
- disable verbose output from pytest

pshapiro4broad committed:
Merge remote-tracking branch 'origin/develop' into ps_DSDEGP-1261_clio_integration_test_in_jenkins_3

pshapiro4broad committed:
DSDEGP-1261: jenkins test cleanups

- eliminate extra sbt project, shouldn't be necessary

pshapiro4broad committed:
DSDEGP-1261: jenkins test cleanups

- also removed jenkins test image, since it's the same as the regular image

pshapiro4broad committed:
DSDEGP-1261: jenkins test cleanups

- merge jenkins test code into other test code
- remove unused files, configs
- disable verbose output from pytest

pshapiro4broad committed:
Merge remote-tracking branch 'origin/ps_DSDEGP-1261_clio_integration_test_in_jenkins_3' into ps_DSDEGP-1261_clio_integration_test_in_jenkins_3

danxmoran commented on clio-integration-test/src/test/resources/docker-compose/docker-compose-jenkins.yml:
```diff
@@ -0,0 +1,14 @@
+  clio-server:
+    image: broadinstitute/clio-server:${CLIO_DOCKER_TAG}
+    environment:
+      clio.server.elasticsearch.http-hosts.0.hostname: elasticsearch1.gotc-dev.broadinstitute.org
```
I'm surprised this works with the dots, we had to change it to [pass through JAVA_OPTS](https://github.com/broadinstitute/clio/blob/develop/clio-integration-test/src/test/resources/docker-compose/docker-compose.yml#L12) in the other docker-compose file to get around the dots-in-name problem. Maybe a difference in how shell expansion works between our local Docker versions and whatever runs on Jenkins.

Can you change it to use the JAVA_OPTS style, so we're consistent across both setups and guarded against the shell behavior changing out from under us?

pshapiro4broad replied:
This worked for me both locally and on the server, before I change it let's try to figure out why it's working for me, and maybe see if it can work for others too.

danxmoran replied:
Want to huddle after the DSDE meetup? I'm curious to get to the bottom of this too.

danxmoran commented on project/src/main/scala/org/broadinstitute/clio/integrationtest/ClioIntegrationTestRunner.scala:
```diff
@@ -76,6 +78,35 @@ class ClioIntegrationTestRunner(testClassesDirectory: File,
     * @return The exit code of the integration test.
     */
+  private def tryDockerComposeJenkinsTest(): Try[Int] = Try {
+    runCommandAsync(
```
I think we can make this work with a single docker-compose call:
```
$ docker-compose --file docker-compose-jenkins.yml up
```
You'll need to update the compose file to declare that the integration-test container depends on the clio container, so it gets the setup ordering right.

danxmoran replied:
^ our deploy config uses this pattern, so it's a good example

pshapiro4broad replied:
Using `up` doesn't work here, we can talk in person about this more.

danxmoran replied:
Thanks for the in-person explanation. Can you add a comment explaining why we have to do it this way?

danxmoran commented on project/src/main/scala/org/broadinstitute/clio/integrationtest/ClioIntegrationTestRunner.scala:
```diff
@@ -132,6 +163,25 @@ class ClioIntegrationTestRunner(testClassesDirectory: File,
+    val processLogger = ProcessLogger(logFunction)
+    process run processLogger
+    ()
+  }
```
If we do keep separate docker-compose calls, I think it'd be better to have this function return a `Future` instead of `Unit`. That way we can detect when the command fails, vs this which is more fire-and-forget.

danxmoran replied:
From in-person: returning `Process` is good too

danxmoran commented on project/src/main/scala/org/broadinstitute/clio/integrationtest/ClioIntegrationTestSettings.scala:
```diff
@@ -13,20 +13,30 @@ object ClioIntegrationTestSettings extends AutoPlugin {
+                                    clioVersion,
+                                    jenkinsTest,
+                                    log)
+        .run()
```
`scalafmt` why

danxmoran commented on project/src/main/scala/org/broadinstitute/clio/sbt/Docker.scala:
```diff
@@ -58,7 +58,7 @@ object Docker {
       copy(pythonSource / "requirements.txt", "./")
       run("pip", "install", "-r", "requirements.txt")
-      entryPoint("pytest", "-vv", "-s")
+      entryPoint("pytest", "-s")
```
Hooray for less noisy output :tada:

danxmoran commented on clio-integration-test/src/test/resources/docker-compose/docker-compose-jenkins.yml:
```diff
@@ -0,0 +1,14 @@
+      ELASTICSEARCH_HTTP_HOST: elasticsearch1.gotc-dev.broadinstitute.org
+    network_mode: "host"
+  clio-server:
+    image: broadinstitute/clio-server:${CLIO_DOCKER_TAG}
```
Rather than declaring all the parameters for the server container here in addition to in the other docker-compose file, could we bust out the definition to a separate YAML and then import it here [like we do for elasticsearch](https://github.com/broadinstitute/clio/blob/develop/clio-integration-test/src/test/resources/docker-compose/docker-compose.yml#L15)? Same for the integration-test-runner container too.

pshapiro4broad replied:
The configs are different so I don't think that will work. E.g. this integration-test service has a different ELASTICSEARCH_HTTP_HOST and requires `network_mode: "host"`.

danxmoran reviewed (changes requested):
Some requests / suggestions! Also a higher-level question for my own understanding: Why not also hit the clio-server running in dev in addition to the elasticsearch cluster running there, rather than spin up a local clio-server container?

pshapiro4broad commented:
> Why not also hit the clio-server running in dev in addition to the elasticsearch cluster running there, rather than spin up a local clio-server container?

This way we can test clio on a branch without having to deploy it first. This seems useful to me but maybe in practice we don't need to do this?

danxmoran commented:
> This seems useful to me but maybe in practice we don't need to do this?

They both seem useful to me :smile: When I think of testing a branch I just think of running everything in isolation, vs. running in an environment when I'd want as many components as possible to be running in that environment. Getting the tests to work against both local and deployed Clio will be pretty complicated because of the authentication layer on the deployed machines, so no worries for now.

danxmoran reviewed (commented):
Deleted comments made outdated by the huddle. Thanks for talking me through it!

pshapiro4broad committed:
DSDEGP-1261: code review feedback

- return a useful value from runCommandAsync
- use JAVA_OPTS instead of passing an environment variable directly to the clio-server docker container

danxmoran reviewed (approved):
Works for me locally now :shipit:

pshapiro4broad committed:
DSDEGP-1261: code review feedback: add comments

pshapiro4broad committed:
Merge branch 'develop' into ps_DSDEGP-1261_clio_integration_test_in_jenkins_3
